lib/rucio/tests/test_reaper2.py

```diff
@@ -203,3 +209,130 @@ def test_reaper_multi_vo(vo):
+    rse_core.set_rse_usage(rse_id=rse_id, source='storage', used=2 * archive_size + nb_c_outside_archive * constituent_size, free=1)
+    reaper(once=True, rses=[], include_rses=rse_name, exclude_rses=None)
+    for did in constituents + [archive1, archive2]:
+        assert did_core.get_did(**did)
+    for did in [archive1, archive2, c_with_replica, c_with_replica_and_rule]:
+        assert replica_core.get_replica(rse_id=rse_id, **did)
+    with pytest.raises(ReplicaNotFound):
+        # The replica is only on the archive, not on the constituent
+        replica_core.get_replica(rse_id=rse_id, **c_first_archive_only)
+    with pytest.raises(ReplicaNotFound):
+        # The replica outside the archive was removed by reaper
+        nb_c_outside_archive -= 1
+        replica_core.get_replica(rse_id=rse_id, **c_with_expired_replica)
+    # Compared to get_replica, list_replicas resolves archives, must return replicas for all files
+    assert len(list(replica_core.list_replicas(dids=constituents))) == 4
+    assert len(list(did_core.list_content(**dataset1))) == 4
+    assert len(list(did_core.list_archive_content(**archive1))) == 4
+    assert len(list(did_core.list_archive_content(**archive2))) == 3
+    assert __get_archive_contents_history_count(archive1) == 0
+    assert __get_archive_contents_history_count(archive2) == 0
+
+    # Expire the first archive and run reaper again
+    # the archive will be removed; and c_first_archive_only must be removed from datasets
+    # and from the did table.
+    replica_core.set_tombstone(rse_id=rse_id, tombstone=datetime.utcnow() - timedelta(days=1), **archive1)
+    REGION.invalidate()
+    rse_core.set_rse_limits(rse_id=rse_id, name='MinFreeSpace', value=2 * archive_size + nb_c_outside_archive * constituent_size)
+    rse_core.set_rse_usage(rse_id=rse_id, source='storage', used=2 * archive_size + nb_c_outside_archive * constituent_size, free=1)
+    reaper(once=True, rses=[], include_rses=rse_name, exclude_rses=None)
+    with pytest.raises(DataIdentifierNotFound):
+        assert did_core.get_did(**archive1)
+    with pytest.raises(DataIdentifierNotFound):
+        assert did_core.get_did(**c_first_archive_only)
+    assert len(list(replica_core.list_replicas(dids=constituents))) == 3
+    assert len(list(did_core.list_content(**dataset1))) == 3
+    assert len(list(did_core.list_archive_content(**archive1))) == 0
+    assert len(list(did_core.list_archive_content(**archive2))) == 3
+    assert __get_archive_contents_history_count(archive1) == 4
+    assert __get_archive_contents_history_count(archive2) == 0
+
+    # Expire the second archive replica and run reaper another time
+    # c_with_expired_replica is removed because its external replica got removed at previous step
+    # and it exist only inside the archive now.
+    # If not open, Dataset2 will be removed because it will be empty.
```

Comment:
We should remove the condition on the dataset open

Reply:
I believe there is related to #3311 . So should be fixed when working on that issue. For now it works like that.